CODING_CONVENTIONS_C++.md

```diff
@@ -0,0 +1,369 @@
+                  << std::endl;
+    }
+}
+
+void my_class::do_something_else() {
+    switch (default_name[0]) {
+        case 'a':
+            // handle a
+            break;
+        case 'b':
+            // handle b
+            break;
+        default:
+          handle_default();
+    }
+}
+
+} // namespace example
+} // namespace riot
+
+```
+
+## General
+
+- Use 4 spaces per indentation level.
+
+- The maximum number of characters per line is 80.
+
+- No tabs, ever.
+
+- Never use C-style casts.
+
+- Vertical whitespaces separate functions and are not used inside functions:
+  use comments to document logical blocks.
+
+- Header filenames end in `.hpp`, implementation files end in `.cpp`.
+
+- Never declare more than one variable per line.
+
+- `*` and `&` bind to the *type*, e.g., `const std::string& arg`.
+
+- Namespaces and access modifiers (e.g., `public`) do not increase the
+  indentation level.
+
+- In a class, use the order `public`, `protected`, and then `private`.
+
+- Always use `auto` to declare a variable unless you cannot initialize it
+  immediately or if you actually want a type conversion. In the latter case,
+  provide a comment why this conversion is necessary.
+
+- Never use unwrapped, manual resource management such as `new` and `delete`.
+
+- Never use `typedef`; always write `using T = X` in favor of `typedef X T`.
+
+- Keywords are always followed by a whitespace: `if (...)`, `template <...>`,
+  `while (...)`, etc.
+
+- Leave a whitespace after `!` to make negations easily recognizable:
+
+  ```cpp
+  if (! sunny())
+      stay_home();
+  else
+      go_outside();
+  ```
+
+- Opening braces belong to the same line:
+
+  ```cpp
+  void foo() {
+      // ...
+  }
+  ```
+
+- Use standard order for readability: C standard libraries, C++ standard
+  libraries, other libraries, (your) RIOT headers:
+
+  ```cpp
+  // some .hpp file
+
+  #include <sys/types.h>
+
+  #include <vector>
+
+  #include "3rd/party.h"
+
+  #include "riot/fwd.hpp"
+  ```
+
+  RIOT includes should always be in doublequotes, whereas system-wide
+  includes in angle brackets. In a `.cpp` file, the implemented header always
+  comes first and the header `riot/config.hpp` can be included second if you
+  need platform-dependent headers.
+
+- When declaring a function, the order of parameters is: outputs, then inputs.
+  This follows the parameter order from the STL.
+
+- Protect single-argument constructors with `explicit` to avoid implicit conversions.
+
+## Naming
+
+- All names except macros and template parameters should be
+  lower case and delimited by underscores.
+
+- Template parameter names should be written in CamelCase.
+
+- Types and variables should be nouns, while functions performing an action
+  should be "command" verbs. Classes used to implement metaprogramming
+  functions also should use verbs, e.g., `remove_const`.
+
+- Private and protected member variables use the suffix `_` while getter *and* 
```

Comment:
Please remove the trailing whitespace here to make Travis happy.

Reply:
Done and squashed